File: TidbitsTrivia/Core/Models/StandingsOutcome.swift
import Foundation

/// How a finished board is announced.
///
/// Every scoreboard in the app used to sort by score and call element 0 "the
/// winner". On a tie that reports an arbitrary sort order as a victory — and
/// ties are not an edge case here: Pass & Play deals ONE shared question set,
/// so two players who answer identically genuinely finish level, and a hosted
/// Trivia Night settles real stakes in a room where a wrong call is visible.
///
/// One helper so all six surfaces (iOS/tvOS/macOS Night + Pass & Play) agree;
/// `js/night.js` and `Party.kt`/`NightHost.kt` mirror the same rule.
nonisolated enum StandingsOutcome {

    /// Every entry sharing the top score. Empty when `entries` is empty.
    static func winners(_ entries: [(name: String, score: Int)]) -> [String] {
        guard let top = entries.map(\.score).max() else { return [] }
        return entries.filter { $0.score == top }.map(\.name)
    }

    /// True when `score` ties the leader — use it to highlight EVERY leading
    /// row, not just the first one the sort happened to put on top.
    static func isTop(_ score: Int, in entries: [(name: String, score: Int)]) -> Bool {
        guard let top = entries.map(\.score).max() else { return false }
        return score == top
    }

    /// The announcement. `empty` is the no-players copy the caller already uses
    /// ("That's a night!"), so this drops into existing sites unchanged.
    static func headline(_ entries: [(name: String, score: Int)], empty: String) -> String {
        let won = winners(entries)
        if won.isEmpty { return empty }
        if won.count == 1 { return "\(won[0]) wins!" }
        if won.count == entries.count { return "It's a tie!" }
        return "Tie — \(won.joined(separator: " & "))"
    }
}

File: windows/Tidbits.Core/Store/StandingsOutcome.cs
namespace Tidbits.Core.Store;

/// <summary>
/// How a finished board is announced.
///
/// Every scoreboard used to sort by score and treat element 0 as "the winner".
/// On a tie that reports an arbitrary sort order as a victory — and ties are not
/// an edge case: Pass &amp; Play deals ONE shared question set, so two players who
/// answer identically genuinely finish level.
///
/// Mirrors Swift <c>Core/Models/StandingsOutcome.swift</c> and the JS/Kotlin
/// copies; kept in Core (not the view) so the rule is a pure, testable function.
/// </summary>
public static class StandingsOutcome
{
    /// <summary>Every name sharing the top score; empty when there are no entries.</summary>
    public static IReadOnlyList<string> Winners(IReadOnlyList<(string Name, int Score)> entries)
    {
        if (entries.Count == 0) return Array.Empty<string>();
        int top = entries.Max(e => e.Score);
        return entries.Where(e => e.Score == top).Select(e => e.Name).ToList();
    }

    /// <summary>True when <paramref name="score"/> ties the leader — highlight EVERY leading row.</summary>
    public static bool IsTop(int score, IReadOnlyList<(string Name, int Score)> entries)
        => entries.Count != 0 && score == entries.Max(e => e.Score);

    /// <summary>The announcement. <paramref name="empty"/> is the no-players copy.</summary>
    public static string Headline(IReadOnlyList<(string Name, int Score)> entries, string empty)
    {
        var won = Winners(entries);
        if (won.Count == 0) return empty;
        if (won.Count == 1) return $"{won[0]} wins!";
        if (won.Count == entries.Count) return "It's a tie!";
        return $"Tie — {string.Join(" & ", won)}";
    }
}
